File: packages/cli/src/translate.js
/**
 * Docker -> wslc translation, driven by the shared rules.json.
 * Keep behaviour in lockstep with packages/py/src/docker2wslc/translate.py.
 */

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
export const RULES = require('../rules.json');

const WINDOWS_PATH = /^[A-Za-z]:[\\/]/;

// Verbs where wslc also accepts the docker-style top-level form, so rewriting
// to the noun form is cosmetic rather than required. Verified on wslc 2.9.4.
const ALIAS_OK = new Set(Object.keys((RULES.renamedOptional || {}).commands || {}));
const SEV_ORDER = { info: 0, warn: 1, error: 2 };

export function tokenize(line) {
  const out = [];
  let cur = '';
  let quote = null;
  for (const ch of line) {
    if (quote) {
      cur += ch;
      if (ch === quote) quote = null;
    } else if (ch === '"' || ch === "'") {
      quote = ch;
      cur += ch;
    } else if (/\s/.test(ch)) {
      if (cur) { out.push(cur); cur = ''; }
    } else {
      cur += ch;
    }
  }
  if (cur) out.push(cur);
  return out;
}

function flagSpec(flag) {
  const flags = RULES.flags;
  if (flags[flag]) return [flag, flags[flag]];
  for (const [canonical, spec] of Object.entries(flags)) {
    if ((spec.aliases || []).includes(flag)) return [canonical, spec];
  }
  return [null, null];
}

function translateArgs(args, notes) {
  const out = [];
  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    const eq = arg.indexOf('=');
    const bare = eq > -1 ? arg.slice(0, eq) : arg;
    const inline = eq > -1 ? arg.slice(eq + 1) : null;
    const [, spec] = flagSpec(bare);

    if (!spec) { out.push(arg); i += 1; continue; }

    const takesValue = Boolean(spec.takesValue);
    const hasInline = eq > -1;
    const value = hasInline ? inline : (args[i + 1] ?? '');
    const consumed = hasInline || !takesValue ? 1 : 2;

    if (spec.action === 'conditional') {
      const branch = (spec.whenValue || {})[value];
      if (branch && branch.action === 'drop') {
        notes.push({ severity: branch.severity || 'info', text: branch.note });
        i += consumed;
        continue;
      }
      if (branch && branch.action === 'error') {
        // Value is kept so the reader sees what they wrote, but flagged as an
        // error: wslc rejects it outright.
        notes.push({ severity: branch.severity || 'error', text: branch.note });
        out.push(hasInline ? `${bare}=${value}` : bare);
        if (!hasInline && takesValue && value) out.push(value);
        i += consumed;
        continue;
      }
      notes.push({ severity: spec.severity || 'info', text: spec.note });
      if (hasInline) out.push(`${bare}=${value}`);
      else { out.push(bare); if (takesValue && value) out.push(value); }
      i += consumed;
      continue;
    }

    if (spec.action === 'drop') {
      notes.push({ severity: spec.severity || 'warn', text: spec.note });
      i += consumed;
      continue;
    }

    if (spec.action === 'rewrite') {
      notes.push({ severity: spec.severity || 'info', text: spec.note });
      out.push(...spec.replaceWith.split(' '));
      i += consumed;
      continue;
    }

    // keep
    if (spec.noteWhenWindowsPath && (WINDOWS_PATH.test(value) || value.startsWith('/mnt/'))) {
      notes.push({ severity: spec.severity || 'info', text: spec.noteWhenWindowsPath });
    }
    // Some flags exist but reject certain value formats (verified: wslc wants
    // uppercase size units, so `-m 512m` fails while `512M` works).
    if (spec.requireValuePattern && value && !new RegExp(spec.requireValuePattern).test(value)) {
      notes.push({ severity: 'error', text: spec.noteWhenValueRejected });
    }
    if (spec.noteAlways) {
      notes.push({ severity: spec.severity || 'info', text: spec.noteAlways });
    }
    out.push(arg);
    if (!hasInline && takesValue && value) out.push(value);
    i += consumed;
  }
  return out;
}

export function exitCodeFor({ notes, composeHit, unsupportedHit }) {
  if (unsupportedHit || composeHit) return 2;
  if (notes.some((n) => n.severity === 'warn' || n.severity === 'error')) return 1;
  return 0;
}

export function translateLine(raw) {
  const line = raw.trim();
  if (!line) return null;
  if (line.startsWith('#')) return { output: line, notes: [] };

  const notes = [];
  const tokens = tokenize(line);

  if (tokens[0] === 'sudo') {
    tokens.shift();
    notes.push({
      severity: 'info',
      text: 'Dropped sudo — wslc runs as your Windows user, no elevation needed for normal container operations.',
    });
  }

  if (!tokens.length || (tokens[0] !== 'docker' && tokens[0] !== 'podman')) {
    return {
      output: `# not a docker command: ${line}`,
      notes: [{ severity: 'info', text: 'Only docker/podman commands are translated. Line left unchanged.' }],
    };
  }

  if (tokens[0] === 'podman') {
    notes.push({ severity: 'info', text: 'Treated podman as Docker-compatible; flag coverage is nearly identical.' });
  }
  tokens.shift();

  if (!tokens.length) return { output: 'wslc', notes };

  if (tokens[0] === 'compose' || tokens[0] === 'docker-compose') {
    notes.push({ severity: 'error', text: RULES.compose.note });
    return { output: '# No native Compose runtime in wslc.', notes, composeHit: true };
  }

  const two = tokens.length > 1 ? `${tokens[0]} ${tokens[1]}` : '';
  // Two-word forms with a verified verdict of their own (e.g. `docker system
  // prune` has no wslc equivalent, but the per-noun prunes do).
  const twoWord = RULES.twoWord || {};
  if (two && twoWord[two]) {
    const entry = twoWord[two];
    if (entry.status === 'unsupported') {
      notes.push({ severity: 'error', text: entry.note });
      return { output: `# ${two}: not supported by wslc`, notes, unsupportedHit: true };
    }
    notes.push({ severity: 'info', text: entry.note });
    const rest = translateArgs(tokens.slice(2), notes);
    return { output: ['wslc', two, ...rest].join(' ').trim(), notes };
  }
  if (two && Object.values(RULES.renamed).includes(two)) {
    const rest = translateArgs(tokens.slice(2), notes);
    return { output: ['wslc', two, ...rest].join(' ').trim(), notes };
  }

  const verb = tokens[0];

  if (RULES.unsupported[verb]) {
    notes.push({ severity: 'error', text: RULES.unsupported[verb] });
    return { output: `# ${verb}: not supported by wslc`, notes, unsupportedHit: true };
  }

  if (RULES.renamed[verb]) {
    const mapped = RULES.renamed[verb];
    const flagMap = RULES.renamedFlags[verb] || {};
    const rest = translateArgs(tokens.slice(1).map((a) => flagMap[a] || a), notes);
    notes.push({
      severity: 'info',
      text: `\`docker ${verb}\` is grouped under a noun in wslc: \`wslc ${mapped}\`. ` + (ALIAS_OK.has(verb)
        ? `\`wslc ${verb}\` also works as a top-level alias (verified on 2.9.4), so this rewrite is cosmetic.`
        : `There is no top-level \`wslc ${verb}\`; the noun form is required.`),
    });
    return { output: ['wslc', mapped, ...rest].join(' ').trim(), notes };
  }

  if (RULES.identical.includes(verb)) {
    const rest = translateArgs(tokens.slice(1), notes);
    return { output: ['wslc', verb, ...rest].join(' ').trim(), notes };
  }

  notes.push({
    severity: 'warn',
    text: `Unknown docker subcommand \`${verb}\` — passed through unchanged. Verify against \`wslc --help\`.`,
  });
  const rest = translateArgs(tokens.slice(1), notes);
  return { output: ['wslc', verb, ...rest].join(' ').trim(), notes };
}

export function translate(text) {
  const lines = [];
  const notes = [];
  const seen = new Set();
  let composeHit = false;
  let unsupportedHit = false;

  for (const raw of String(text).split('\n')) {
    const res = translateLine(raw);
    if (!res) { lines.push(''); continue; }
    lines.push(res.output);
    composeHit = composeHit || Boolean(res.composeHit);
    unsupportedHit = unsupportedHit || Boolean(res.unsupportedHit);
    for (const note of res.notes) {
      if (!seen.has(note.text)) { seen.add(note.text); notes.push(note); }
    }
  }

  notes.sort((a, b) => (SEV_ORDER[b.severity] ?? 0) - (SEV_ORDER[a.severity] ?? 0));
  const result = {
    output: lines.join('\n').replace(/\n{3,}/g, '\n\n').trim(),
    notes,
    composeHit,
    unsupportedHit,
  };
  result.exitCode = exitCodeFor(result);
  return result;
}

File: packages/py/src/docker2wslc/translate.py
"""Docker -> wslc command translation, driven by the shared rules.json."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from importlib.resources import files
from typing import Any

WINDOWS_PATH = re.compile(r"^[A-Za-z]:[\\/]")


def load_rules() -> dict[str, Any]:
    """Load the bundled rules.json (single source of truth across languages)."""
    data = files("docker2wslc").joinpath("rules.json").read_text(encoding="utf-8")
    return json.loads(data)


RULES = load_rules()

SEV_ORDER = {"info": 0, "warn": 1, "error": 2}


@dataclass
class Note:
    severity: str
    text: str

    def __str__(self) -> str:  # pragma: no cover - display helper
        return f"[{self.severity}] {self.text}"


@dataclass
class Result:
    """Outcome of translating one or more lines."""

    output: str = ""
    notes: list[Note] = field(default_factory=list)
    compose_hit: bool = False
    unsupported_hit: bool = False

    @property
    def exit_code(self) -> int:
        """0 clean, 1 degraded (flags dropped/rewritten), 2 unmigratable."""
        if self.unsupported_hit or self.compose_hit:
            return 2
        if any(n.severity in ("warn", "error") for n in self.notes):
            return 1
        return 0

    def as_dict(self) -> dict[str, Any]:
        return {
            "output": self.output,
            "notes": [{"severity": n.severity, "text": n.text} for n in self.notes],
            "composeHit": self.compose_hit,
            "unsupportedHit": self.unsupported_hit,
            "exitCode": self.exit_code,
        }


def tokenize(line: str) -> list[str]:
    """Split a shell-ish line, keeping quoted spans intact (quotes preserved)."""
    out: list[str] = []
    cur = ""
    quote: str | None = None
    for ch in line:
        if quote:
            cur += ch
            if ch == quote:
                quote = None
        elif ch in ("'", '"'):
            quote = ch
            cur += ch
        elif ch.isspace():
            if cur:
                out.append(cur)
                cur = ""
        else:
            cur += ch
    if cur:
        out.append(cur)
    return out


def _flag_spec(flag: str) -> tuple[str, dict[str, Any]] | tuple[None, None]:
    flags = RULES["flags"]
    if flag in flags:
        return flag, flags[flag]
    for canonical, spec in flags.items():
        if flag in spec.get("aliases", []):
            return canonical, spec
    return None, None


def _translate_args(args: list[str], notes: list[Note]) -> list[str]:
    out: list[str] = []
    i = 0
    while i < len(args):
        arg = args[i]
        bare, _, inline = arg.partition("=")
        canonical, spec = _flag_spec(bare)

        if spec is None:
            out.append(arg)
            i += 1
            continue

        takes_value = spec.get("takesValue", False)
        has_inline = "=" in arg
        value = inline if has_inline else (args[i + 1] if i + 1 < len(args) else "")
        consumed = 1 if has_inline or not takes_value else 2

        action = spec["action"]

        if action == "conditional":
            branch = spec.get("whenValue", {}).get(value)
            if branch:
                if branch["action"] == "drop":
                    notes.append(Note(branch.get("severity", "info"), branch["note"]))
                    i += consumed
                    continue
            notes.append(Note(spec.get("severity", "info"), spec["note"]))
            out.append(f"{bare}={value}" if has_inline else bare)
            if not has_inline and takes_value and value:
                out.append(value)
            i += consumed
            continue

        if action == "drop":
            notes.append(Note(spec.get("severity", "warn"), spec["note"]))
            i += consumed
            continue

        if action == "rewrite":
            notes.append(Note(spec.get("severity", "info"), spec["note"]))
            out.extend(spec["replaceWith"].split())
            i += consumed
            continue

        # action == "keep"
        note_win = spec.get("noteWhenWindowsPath")
        if note_win and (WINDOWS_PATH.match(value) or value.startswith("/mnt/")):
            notes.append(Note(spec.get("severity", "info"), note_win))
        out.append(arg)
        if not has_inline and takes_value and value:
            out.append(value)
        i += consumed
    return out


def translate_line(raw: str) -> Result | None:
    """Translate a single line. Returns None for blank lines."""
    line = raw.strip()
    if not line:
        return None
    if line.startswith("#"):
        return Result(output=line)

    notes: list[Note] = []
    tokens = tokenize(line)

    if tokens and tokens[0] == "sudo":
        tokens.pop(0)
        notes.append(
            Note(
                "info",
                "Dropped sudo — wslc runs as your Windows user, no elevation needed "
                "for normal container operations.",
            )
        )

    if not tokens or tokens[0] not in ("docker", "podman"):
        return Result(
            output=f"# not a docker command: {line}",
            notes=[Note("info", "Only docker/podman commands are translated. Line left unchanged.")],
        )

    if tokens[0] == "podman":
        notes.append(
            Note("info", "Treated podman as Docker-compatible; flag coverage is nearly identical.")
        )
    tokens.pop(0)

    if not tokens:
        return Result(output="wslc", notes=notes)

    if tokens[0] in ("compose", "docker-compose"):
        compose = RULES["compose"]
        notes.append(Note("error", compose["note"]))
        return Result(
            output="# No native Compose runtime in wslc.", notes=notes, compose_hit=True
        )

    two = f"{tokens[0]} {tokens[1]}" if len(tokens) > 1 else ""
    if two and two in (RULES["renamed"].values()):
        rest = _translate_args(tokens[2:], notes)
        return Result(output=" ".join(["wslc", two, *rest]).strip(), notes=notes)

    verb = tokens[0]

    if verb in RULES["unsupported"]:
        notes.append(Note("error", RULES["unsupported"][verb]))
        return Result(
            output=f"# {verb}: not supported by wslc", notes=notes, unsupported_hit=True
        )

    if verb in RULES["renamed"]:
        mapped = RULES["renamed"][verb]
        flag_map = RULES["renamedFlags"].get(verb, {})
        rest: list[str] = []
        for arg in tokens[1:]:
            rest.append(flag_map.get(arg, arg))
        rest = _translate_args(rest, notes)
        notes.append(
            Note(
                "info",
                f"`docker {verb}` is grouped under a noun in wslc: `wslc {mapped}`. "
                f"Recent previews accept `wslc {verb}` as an alias, but the noun form is stable.",
            )
        )
        return Result(output=" ".join(["wslc", mapped, *rest]).strip(), notes=notes)

    if verb in RULES["identical"]:
        rest = _translate_args(tokens[1:], notes)
        return Result(output=" ".join(["wslc", verb, *rest]).strip(), notes=notes)

    notes.append(
        Note(
            "warn",
            f"Unknown docker subcommand `{verb}` — passed through unchanged. "
            "Verify against `wslc --help`.",
        )
    )
    rest = _translate_args(tokens[1:], notes)
    return Result(output=" ".join(["wslc", verb, *rest]).strip(), notes=notes)


def translate(text: str) -> Result:
    """Translate a multi-line script, de-duplicating notes."""
    lines: list[str] = []
    notes: list[Note] = []
    seen: set[str] = set()
    compose_hit = False
    unsupported_hit = False

    for raw in str(text).split("\n"):
        res = translate_line(raw)
        if res is None:
            lines.append("")
            continue
        lines.append(res.output)
        compose_hit = compose_hit or res.compose_hit
        unsupported_hit = unsupported_hit or res.unsupported_hit
        for note in res.notes:
            if note.text not in seen:
                seen.add(note.text)
                notes.append(note)

    joined = re.sub(r"\n{3,}", "\n\n", "\n".join(lines)).strip()
    notes.sort(key=lambda n: -SEV_ORDER.get(n.severity, 0))
    return Result(
        output=joined,
        notes=notes,
        compose_hit=compose_hit,
        unsupported_hit=unsupported_hit,
    )
